third_party/vcluster/vcluster.go

```diff
@@ -0,0 +1,324 @@
+)
+
+var vclusterVersion = "v0.20.0"
+
+type Cluster struct {
+	path            string
+	name            string
+	kubecfgFile     string // kubeconfig file for the vcluster
+	version         string
+	namespace       string // namespace to create the vcluster in
+	hostKubeCfg     string // kubeconfig file for the host cluster
+	hostKubeContext string // kubeconfig context for the host cluster
+	rc              *rest.Config
+}
+
+// Enforce Type check always to avoid future breaks
+var _ support.E2EClusterProvider = &Cluster{}
+
+func NewCluster(name string) *Cluster {
+	return &Cluster{name: name}
+}
+
+func NewProvider() support.E2EClusterProvider {
+	return &Cluster{}
+}
+
+func WithPath(path string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.path = path
+		}
+	}
+}
+
+func WithNamespace(ns string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.namespace = ns
+		}
+	}
+}
+
+func WithHostKubeConfig(kubeconfig string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.hostKubeCfg = kubeconfig
+		}
+	}
+}
+
+func WithHostKubeContext(context string) support.ClusterOpts {
+	return func(c support.E2EClusterProvider) {
+		v, ok := c.(*Cluster)
+		if ok {
+			v.hostKubeContext = context
+		}
+	}
+}
+
+func (c *Cluster) WithName(name string) support.E2EClusterProvider {
+	c.name = name
+	return c
+}
+
+func (c *Cluster) WithVersion(version string) support.E2EClusterProvider {
+	c.version = version
+	return c
+}
+
+func (c *Cluster) WithPath(path string) support.E2EClusterProvider {
+	c.path = path
+	return c
+}
+func (c *Cluster) WithOpts(opts ...support.ClusterOpts) support.E2EClusterProvider {
+	for _, opt := range opts {
+		opt(c)
+	}
+	return c
+}
+func (c *Cluster) SetDefaults() support.E2EClusterProvider {
+	if c.path == "" {
+		c.path = "vcluster"
+	}
+	return c
+}
+func (c *Cluster) Create(ctx context.Context, args ...string) (string, error) {
+	log.V(4).Info("Creating vcluster ", c.name)
+	if err := c.findOrInstallVcluster(); err != nil {
+		return "", err
+	}
+
+	if _, exists := c.clusterExists(c.name); exists {
+		log.V(4).Info("Skipping vcluster Cluster.Create: cluster already created: ", c.name)
+		kConfig, err := c.getKubeconfig()
+		if err != nil {
+			return "", err
+		}
+		return kConfig, c.initKubernetesAccessClients()
+	}
+
+	if c.namespace != "" {
+		args = append(args, "--namespace", c.namespace)
+	}
+
+	if c.hostKubeContext != "" {
+		args = append(args, "--context", c.hostKubeContext)
+	}
+
+	command := fmt.Sprintf("%s create %s --connect=false --update-current=false", c.path, c.name)
+	if len(args) > 0 {
+		command = fmt.Sprintf("%s %s", command, strings.Join(args, " "))
+	}
+	log.V(4).Info("Launching:", command)
+	echo := gexe.New()
+	if c.hostKubeCfg != "" {
+		echo.SetEnv("KUBECONFIG", c.hostKubeCfg)
+	}
+
+	p := echo.RunProc(command)
```

Comment:
I think @harshanarayana has some util wrapper for running procs that should be considered here.